### PR TITLE
🚀 Anomalib Pipelines 

### PR DESCRIPTION
**Pull request by ashwinvaidya17**
## 📝 Description

- Alternate design for pipeline

## ✨ Changes

Select what type of change your PR is:

- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [ ] 🔨 Refactor (non-breaking change which refactors the code base)
- [x] 🚀 New feature (non-breaking change which adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📚 Documentation update
- [ ] 🔒 Security update

## ✅ Checklist

Before you submit your pull request, please make sure you have completed the following steps:

- [ ] 📋 I have summarized my changes in the [CHANGELOG](https://github.com/openvinotoolkit/anomalib/blob/main/CHANGELOG.md) and followed the guidelines for my type of change (skip for minor changes, documentation updates, and test enhancements).
- [ ] 📚 I have made the necessary updates to the documentation (if applicable).
- [ ] 🧪 I have written tests that support my changes and prove that my fix is effective or my feature works (if applicable).

For more information about code review checklists, see the [Code Review Checklist](https://github.com/openvinotoolkit/anomalib/blob/main/docs/source/markdown/guides/developer/code_review_checklist.md).

### CODE REVIEW

File: src/anomalib/pipelines/benchmark/job.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    def save(result: pd.DataFrame) -> None:
+        """Save the result to a csv file."""
+        BenchmarkJob._print_tabular_results(result)
+        file_path = Path("runs") / BenchmarkJob.name / datetime.now().strftime("%Y-%m-%d-%H:%M:%S") / "results.csv"
+        file_path.parent.mkdir(parents=True, exist_ok=True)
+        result.to_csv(file_path, index=False)
+        logger.info(f"Saved results to {file_path}")
+
+    @staticmethod
+    def _print_tabular_results(gathered_result: pd.DataFrame) -> None:
```
Comment (samet-akcay) [inline src/anomalib/pipelines/benchmark/job.py:91]: would this be used by some other jobs or some components in anomalib? If so, would it be an idea to move this to a utility module?

File: src/anomalib/pipelines/benchmark/pipeline.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+from jsonargparse import ArgumentParser, Namespace
+
+from anomalib.pipelines.components.base import Pipeline, Runner
+from anomalib.pipelines.components.runners import ParallelRunner, SerialRunner
+
+from .generator import BenchmarkJobGenerator
+
+
+class Benchmark(Pipeline):
+    """Benchmarking orchestrator."""
```
Comment (samet-akcay) [inline src/anomalib/pipelines/benchmark/pipeline.py:16]: Are we still using `orchestrator`? Or is it a left-over?

File: src/anomalib/pipelines/benchmark/pipeline.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        """Setup the runners for the pipeline."""
+        accelerators = args.accelerator if isinstance(args.accelerator, list) else [args.accelerator]
+        runners: list[Runner] = []
+        for accelerator in accelerators:
+            if accelerator == "cpu":
+                runners.append(SerialRunner(BenchmarkJobGenerator("cpu")))
+            elif accelerator == "cuda":
+                runners.append(
+                    ParallelRunner(BenchmarkJobGenerator("cuda"), n_jobs=torch.cuda.device_count()),
+                )
```
Comment (samet-akcay) [inline src/anomalib/pipelines/benchmark/pipeline.py:28]: This seems to be fitting to a single line. Do you intentionally break it to multiple lines?

File: src/anomalib/pipelines/benchmark/pipeline.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    def get_parser(self, parser: ArgumentParser | None = None) -> ArgumentParser:
+        """Add arguments to the parser."""
+        parser = super().get_parser(parser)
+        parser.add_argument(
+            "--accelerator",
+            type=str | list[str],
+            default="cuda",
+            help="Hardware to run the benchmark on.",
+        )
+        BenchmarkJobGenerator.add_arguments(parser)
```
Comment (samet-akcay) [inline src/anomalib/pipelines/benchmark/pipeline.py:43]: Would an implementer always need to pass the parser to `JobGenerator`? or is this implementation-specific ?

File: src/anomalib/pipelines/components/actions/grid_search.py [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+
+from jsonargparse import Namespace
+from jsonargparse._typehints import ActionTypeHint
+
+from anomalib.pipelines.components import (
+    convert_to_tuple,
+    dict_from_namespace,
+    flatten_dict,
+    namespace_from_dict,
+    to_nested_dict,
```
Comment (samet-akcay) [inline src/anomalib/pipelines/components/actions/grid_search.py:19]: are these pipeline specific implementations? could they go somewhere in `anomalib.utils`. I remember there are some similar util functions somewhere else in the repo. nncf maybe

File: src/anomalib/pipelines/components/actions/grid_search.py [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+            _container[key.removesuffix(".grid")] = value
+        yield to_nested_dict(_container)
+
+
+class GridSearchAction(ActionTypeHint):
+    """Grid search action for jsonargparse.
+
+    Allows using grid search key in the arguments.
+
+    Example:
```
Comment (samet-akcay) [inline src/anomalib/pipelines/components/actions/grid_search.py:69]:
```suggestion
    Examples:
```

File: src/anomalib/pipelines/components/actions/grid_search.py [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+        ```yaml
+
+            nested_key:
+                class_path: str
+                init_args: [arg1, arg2]
+        ```
+        or
```
Comment (samet-akcay) [inline src/anomalib/pipelines/components/actions/grid_search.py:76]:
```suggestion
        
        or
```

File: src/anomalib/pipelines/components/actions/grid_search.py [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+        """Returns a new value with all grid search keys removed."""
+        _value = dict_from_namespace(value) if isinstance(value, Namespace) else copy.deepcopy(value)
+        keys: list[str] = []
+        GridSearchAction._crawl(_value, keys)
+        for key in keys:
+            GridSearchAction._pop_nested_key(_value, key)
+        return _value
+
+    def _check_type(self, value: dict | Namespace | Any, append: bool = False, cfg: Namespace | None = None) -> Any:  # noqa: ANN401
+        """Ignore all grid search keys.
```
Comment (samet-akcay) [inline src/anomalib/pipelines/components/actions/grid_search.py:132]: Not sure if these docstrings are sphinx friendly. can you double check these. For example yaml examples could be done as follows: 

.. code-block:: yaml
    ....

File: src/anomalib/pipelines/components/actions/grid_search.py [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+
+        This allows the args to follow the same format as ``add_subclass_arguments``
+        nested_key:
+            class_path: str
+            init_args: [arg1, arg2]
```
Comment (samet-akcay) [inline src/anomalib/pipelines/components/actions/grid_search.py:137]: Something like this
```suggestion
        .. code-block:: yaml

            nested_key:
                class_path:
                    grid: [val1, val2]  # List of values to grid search
                init_args:
                    - arg1
                    - arg2
```

File: src/anomalib/pipelines/components/base/pipeline.py [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+log_file = "runs/pipeline.log"
+Path(log_file).parent.mkdir(exist_ok=True, parents=True)
+logger_file_handler = logging.FileHandler(log_file)
+logging.getLogger().addHandler(logger_file_handler)
+logging.getLogger().setLevel(logging.DEBUG)
+warnings.filterwarnings("ignore")
+for logger_name in ["lightning.pytorch", "lightning.fabric", "torchmetrics", "os"]:
+    logging.getLogger(logger_name).handlers = [logger_file_handler]
+format_string = "%(asctime)s - %(name)s - %(levelname)s - %(message)s"
+logging.basicConfig(format=format_string, level=logging.DEBUG)
```
Comment (samet-akcay) [inline src/anomalib/pipelines/components/base/pipeline.py:24]: Would it be possible to wrap this in a function, like `setup_logging` or something similar. It looks a bit messy this way

File: tests/integration/pipelines/__init__.py [hunk 11]
```diff
@@ -0,0 +1,5 @@
+"""Pipeline tests."""
+
+
```
Comment (samet-akcay) [inline tests/integration/pipelines/__init__.py:3]:
```suggestion
```

File: tests/integration/pipelines/test_benchmark.py [hunk 12]
```diff
@@ -0,0 +1,20 @@
+"""Test benchmarking pipeline."""
+
+# Copyright (C) 2024 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+from unittest.mock import MagicMock, patch
```
Comment (samet-akcay) [inline tests/integration/pipelines/test_benchmark.py:5]:
```suggestion

from unittest.mock import MagicMock, patch
```

File: tools/experimental/README.md [hunk 13]
```diff
@@ -0,0 +1,3 @@
+# Anomalib Experimental
+
+These are experimental utilities that are under development. These might change frequently or might even be dropped.
```
Comment (samet-akcay) [inline tools/experimental/README.md:3]: can we add this in a warning section

**Review by samet-akcay — CHANGES_REQUESTED** [review]
took me some time, but managed to do my first round :)

It is looking good so far! I've got some minor comments initially. Will go for another round later

File: src/anomalib/pipelines/benchmark/generator.py [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+        for _container in get_iterator_from_grid_dict(container):
+            yield BenchmarkJob(
+                accelerator=self.accelerator,
+                seed=_container["seed"],
+                model=get_model(_container["model"]),
+                datamodule=get_datamodule(_container["data"]),
+            )
+
+    @staticmethod
+    def _add_subclass_arguments(parser: ArgumentParser, baseclass: type, key: str) -> None:
```
[the remark this answers is not in the document]
Comment (ashwinvaidya17) [reply src/anomalib/pipelines/benchmark/generator.py:60]: We can't call the default `add_subclass_arguments` of jsonargparse as we need to override the action. This method passes `GridSearchAction` to action parameter when registering class arguments.

File: src/anomalib/pipelines/benchmark/pipeline.py (hunk 4 above)
Comment (ashwinvaidya17) [reply src/anomalib/pipelines/benchmark/pipeline.py:43]: They have to pass it manually each time as job specific arguments are added to the same parser.

File: src/anomalib/pipelines/benchmark/pipeline.py (hunk 2 above)
Comment (ashwinvaidya17) [reply src/anomalib/pipelines/benchmark/pipeline.py:16]: I missed this

File: src/anomalib/pipelines/benchmark/job.py [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+        self.seed = seed
+
+    @hide_output
+    def run(
+        self,
+        task_id: int | None = None,
+    ) -> dict[str, Any]:
+        """Run the benchmark."""
+        devices: str | list[int] = "auto"
+        if task_id is not None:
```
[the remark this answers is not in the document]
Comment (ashwinvaidya17) [reply src/anomalib/pipelines/benchmark/job.py:44]: This job makes this assumption. The benchmark pipeline sets the process pool to the number of cuda devices.

File: src/anomalib/pipelines/benchmark/job.py (hunk 1 above)
Comment (ashwinvaidya17) [reply src/anomalib/pipelines/benchmark/job.py:88]: Once we add the csv logger we can support printing the results to the console. Maybe then we can make this independent?

**Comment by blaz-r** [comment]
I glanced over the design real quick, seems great 😄. I'll do a more thorough overview asap, mostly from the viewpoint of the tiled ensemble.

File: src/anomalib/pipelines/benchmark/generator.py [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+        """Add job specific arguments to the parser."""
+        group = parser.add_argument_group("Benchmark job specific arguments.")
+        group.add_argument(
+            f"--{BenchmarkJob.name}.seed",
+            type=int | dict[str, list[int]],
+            default=42,
+            help="Seed for reproducibility.",
+        )
+        BenchmarkJobGenerator._add_subclass_arguments(group, AnomalyModule, f"{BenchmarkJob.name}.model")
+        BenchmarkJobGenerator._add_subclass_arguments(group, AnomalibDataModule, f"{BenchmarkJob.name}.data")
```
Comment (djdameln) [inline src/anomalib/pipelines/benchmark/generator.py:44]: I would prefer the use of `self` here for readability
```suggestion
        self._add_subclass_arguments(group, AnomalyModule, f"{BenchmarkJob.name}.model")
        self._add_subclass_arguments(group, AnomalibDataModule, f"{BenchmarkJob.name}.data")
```

**Review by djdameln — COMMENTED** [review]
Thanks, I think we're getting there. I like the new Generator design which returns the job instance. My biggest concern with the current design is that the argument parsing is spread across multiple classes, which makes it a bit hard to follow. This may make it a bit intimidating for users to implement their own custom pipeline. Do you think we could simplify this in some way?

Since we want to encourage users to implement their own custom pipelines, we need to make sure that the functionality of the different components is very clear. I think it would be good to add a bit more detail to the docstrings of the classes (Pipeline, Job, Generator, Runner) including some examples to make it easier for the users to follow.

In line with this, I think it would be good to also add a guide to our documentation explaining step by step how to implement a new pipeline.

File: src/anomalib/pipelines/components/runners/parallel.py [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+    from concurrent.futures import Future
+
+logger = logging.getLogger(__name__)
+
+
+class ParallelExecutionError(Exception):
+    """Pool execution error should be raised when one or more jobs fail in the pool."""
+
+
+class ParallelRunner(Runner):
```
[the remark this answers is not in the document]
Comment (ashwinvaidya17) [reply src/anomalib/pipelines/components/runners/parallel.py:27]: The parallel runner just creates an execution pool and passes the task_id to the job's `run` method. The job is responsible for using this task_id to use the appropriate device.

File: src/anomalib/pipelines/benchmark/generator.py [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+from anomalib.utils.logging import hide_output
+
+from .job import BenchmarkJob
+
+
+class BenchmarkJobGenerator(JobGenerator):
+    """Generate BenchmarkJob."""
+
+    def __init__(self, accelerator: str) -> None:
+        self.accelerator = accelerator
```
[the remark this answers is not in the document]
Comment (ashwinvaidya17) [reply src/anomalib/pipelines/benchmark/generator.py:26]: Originally this was called device. I think we discussed on changing this to accelerator to be inline with lightning's terminology. I have no preference here. So, I can rename this once we finalise the name.

[the remark this answers is not in the document]
Comment (samet-akcay) [reply src/anomalib/pipelines/benchmark/generator.py:21]: In Lightning, accelerator seems to be also used for CPUs
https://lightning.ai/docs/pytorch/stable/extensions/accelerator.html
https://lightning.ai/docs/pytorch/stable/api/lightning.pytorch.accelerators.CPUAccelerator.html

File: tools/experimental/README.md
Comment (samet-akcay) [inline tools/experimental/README.md:1]: Would it be an idea to use `warnings.warn` in the entrypoint scripts to inform the user whenever they use these features?

**Review by samet-akcay — APPROVED** [review]
Looking good to me. My only feedback is the lack of examples in the docstring and documentation. If you would like to add this in a follow-up PR, that is also fine.

**Comment by blaz-r** [comment]
I will check this thoroughly in a few hours and leave some feedback.

**Comment by blaz-r** [comment]
For some reason I keep getting `rich.errors.LiveError: Only one live display may be active at once` when trying to run the benchmark, although I'm only using Padim.

**Comment by blaz-r** [comment]
I checked the code and the design is quite nice. I have just one question that I commented on the relevant part.

File: src/anomalib/pipelines/components/base/pipeline.py [hunk 19]
```diff
@@ -0,0 +1,77 @@
...
+            try:
+                _args = args.get(runner.generator.job_class.name, None)
+                runner.run(_args)
+            except Exception:  # noqa: PERF203 catch all exception and allow try-catch in loop
+                logger.exception("An error occurred when running the runner.")
+                print(
+                    f"There were some errors when running [red]{runner.generator.job_class.name}[/red] with"
+                    f" [green]{runner.__class__.__name__}[/green]."
+                    f" Please check [magenta]{log_file}[/magenta] for more details.",
+                )
```
[the remark this answers is not in the document]
Comment (ashwinvaidya17) [reply src/anomalib/pipelines/components/base/pipeline.py:68]: Then I guess this design falls short 🙂 I'll have a look again at your code to see what changes need to be made.

**Comment by ashwinvaidya17** [comment]
> For some reason I keep getting rich.errors.LiveError: Only one live display may be active at once when trying to run the benchmark, although I'm only using Padim.

I'll have a look. Even patchcore breaks due to rich progress bar in kcenter-greedy. Maybe we shouldn't merge it till this issue is not solved

File: src/anomalib/pipelines/components/base/pipeline.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (blaz-r) [reply src/anomalib/pipelines/components/base/pipeline.py:68]: Yeah. I think that having a pipeline, at least for me, would mean that you chain some elements and result of one is input for the next one. I also see no problem of using pipelines recursively (not a very deep recursion). So let's says in case of the tiled ensemble it'd look something like:
Training[Parallel]->merging[serial]->postprocessing(implemented as sub-pipeline consisting of serialRunners each doing one step like threshold, norm ...).
